mgr: fixup pgs show in unknown state

renhwztetecs opened the pull request:
pgs show the unknown state, when the mgr has just started, use the command ceph pg dump pgs_brief\pgs\...,
because the pgs state has not been reported from the Osd to the Mgr service,
We need to wait for this state before we can run the ceph pg related state on Mgr.

Fixes: http://tracker.ceph.com/issues/25103
Signed-off-by: huanwen ren <ren.huanwen@zte.com.cn>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

renhwztetecs commented:
I am here to choose to exit directly, when "pgmap_ready" is not ready.  Maybe can we consider waiting until "pgmap_ready" is ready? @jcsp

liewegas commented:
Implementation looks fine.  I'm a little unsure whether we want to hide the unknown pgs from the 'pg dump' user or not, though... as a human operator I'd like to see it.

jcsp commented:
Not sure how related this is to the #25103 ticket - @liewegas the observation in that ticket seems to be about certain PGs being unknown when they shouldn't, rather than about whether to show info for unknown pgs, right?

jcsp commented:
Perhaps the right thing to do here is have that "PG state is not ready" message on the stderr output, but then proceed to give the actual state (including unknowns) on stdout.

@renhwztetecs I wonder if there was a specific use case you were thinking of, like a monitoring tool using this command?  Maybe we could also modify the JSON output to have a field that just passes through pgmap_ready

renhwztetecs commented:
@jcsp  We use the 'ceph pg dump pgs_brief' command to get the pg status information in the Calamari/Dashboard, Maybe 'ceph pg dump pgs_brief' has no suitable place to increase the pgmap_ready field?

I don't understand what it means to get the pg status when ‘pgmap_ready’ is not ready. I understand that the information obtained at this time should be inaccurate or incomplete information?@tchaikov

jcsp commented:
> We use the 'ceph pg dump pgs_brief' command to get the pg status information in the Calamari/Dashboard, Maybe 'ceph pg dump pgs_brief' has no suitable place to increase the pgmap_ready field?

Hmm, it is annoying that pgs_brief is an array rather than a dictionary.

Calamari is very old, but we should think about how to handle this for the dashboard.  Maybe for the dashboard it would also be good to do something like greying out the chart until the data is fresh.

So we probably just need to fix the formats of our outputs, so that we can include the pgmap_ready field -- that's easy for the internal interface between ceph-mgr and the dashboard, but a little more awkward for any external tool that was depending on the old pgs_brief format.

> I don't understand what it means to get the pg status when ‘pgmap_ready’ is not ready. I understand that the information obtained at this time should be inaccurate or incomplete information?

Yes, but sometimes the incomplete information is interesting.  For example, if the administrator would like to know which PGs are in an unknown state (OSDs not reporting).

renhwztetecs commented:
Thanks for the reply.
 I removed the forced exit when the pgmap is not ready, just give a hint to judge the Management System.

renhwztetecs commented:
@jcsp  This is my Mistake, you are right, adding to the CLI does not solve the problem.
Added print information to stderr for human readable message, and the value of pgmap_ready is also added to Formatter output.

renhwztetecs commented on src/mgr/DaemonServer.cc:
```diff
@@ -1998,6 +1998,15 @@ bool DaemonServer::handle_command(MCommand *m)
+    if (!pgmap_ready) {
+      derr << "Warning: due to ceph-mgr restart, some PG states may not be up to date" << dendl;
+    }
+    if (f) {
```
JSON prints as follows:

{
    "pgmap_ready": false
}
[
    {
        "pgid": "3.1ff",
        "state": "unknown",
        "up": [],
        "acting": [],
        "up_primary": -1,
        "acting_primary": -1
    },
    ......
]

renhwztetecs replied:
[] exists before modification

renhwztetecs replied:
I will modifying this place.

renhwztetecs replied:
@jcsp thanks
Fixed, JSON prints as follows:
{
    "pgmap_ready": {
        "pgmap_ready": false
    },
    "pg_stats": [
        {
            "pgid": "3.1ff",
            "state": "unknown",
            "up": [],
            "acting": [],
            "up_primary": -1,
            "acting_primary": -1
        },
        ......
    ]
}

jcsp replied:
Sorry about the delay.  This looks good to me!

renhwztetecs commented on src/mgr/DaemonServer.cc:
```diff
@@ -1998,6 +1998,15 @@ bool DaemonServer::handle_command(MCommand *m)
     return true;
   } else {
+    if (!pgmap_ready) {
+      derr << "Warning: due to ceph-mgr restart, some PG states may not be up to date" << dendl;
```
add cli output.

tchaikov replied:
please note, this is not an error in the mgr daemon, so i don't think we should send this error message to its log file.

renhwztetecs replied:
@tchaikov thanks for your reviews.
There is a situation, such as when managing the system timing scheduling, if you do not log to the log, we may not be able to backtrack the positioning problem?

renhwztetecs replied:
Delete input into the err log,  We don't need to pay attention to the log。

tchaikov commented on src/mgr/DaemonServer.cc:
```diff
@@ -1998,6 +1998,17 @@ bool DaemonServer::handle_command(MCommand *m)
   } else {
+    if (!pgmap_ready) {
+      derr << "Warning: due to ceph-mgr restart, some PG states may not be up to date" << dendl;
+      ss << "Warning: due to ceph-mgr restart, some PG states may not be up to date\n";
```
why do you put a newline at the end of the warning message? and i am inclined to not include "Warning:" in this message. as `ss` is supposed to include the error messages.

renhwztetecs replied:
If we do not add "\n", it will become the following print(connected with ‘dumped pgs_brief ’):

>[root@cephfs102 ~]# ceph pg dump pgs_brief    
>Warning: due to ceph-mgr restart, some PG states may not be up to date dumped pgs_brief 

'ss' output to the cli interface, there is no redundant error message identification, such as printing information above, if not, increase the "Warning:" can better prompt the user?

renhwztetecs commented:
Delete input into the err log, We don't need to pay attention to the log.

tchaikov commented:
@renhwztetecs sorry for the latency. i will give this PR a spin locally first.

tchaikov commented:
@renhwztetecs could you take at the failures in http://pulpito.ceph.com/kchai-2018-09-06_14:00:49-rados-wip-kefu-testing-2018-09-06-1316-distro-basic-mira/ 

i think your change is not compatible with what `get_pg_stats()` expects.

renhwztetecs commented:
@tchaikov 
updated, maybe again.
http://pulpito.ceph.com/xxg-2018-09-08_01:48:05-rados-renhw-wip-25103-distro-basic-smithi/

xiexingguo commented:
still causing error, e.g.:

http://qa-proxy.ceph.com/teuthology/xxg-2018-09-08_10:17:50-rados-wip-xxg-testing-2018-09-08-1637-distro-basic-smithi/2994997/teuthology.log

xiexingguo commented:
@renhwztetecs Is this ready for retest?

renhwztetecs commented:
> still causing error, e.g.:
> 
> http://qa-proxy.ceph.com/teuthology/xxg-2018-09-08_10:17:50-rados-wip-xxg-testing-2018-09-08-1637-distro-basic-smithi/2994997/teuthology.log

This is a test using the latest qa to test the old ceph service version, because the new pg dump format is used in the new qa, so this error will be reported, can it be ignored, or consider qa compatibility? Is there information in Qa that determines the cephalosporin version?

In addition to this, others have been prepared.

jcsp commented:
@renhwztetecs it's important that we don't break the upgrade tests.  You can update the places in `qa/` that access pg_stats to accept either the old or the new format (e.g. access `['pg_stats']`, catch `KeyError` and fall back to the old style)

renhwztetecs commented:
@jcsp updated, I added "try except"
I didn't think it was necessary to consider this compatibility before. :-)

tchaikov commented on qa/tasks/ceph_manager.py:
```diff
@@ -1786,7 +1786,13 @@ def get_pg_stats(self):
-        return j['pg_stats']
+        try:
+            new_stats = j['pg_map']['pg_stats']
+        except:
```
please be specific that we are expecting `KeyError` here.

renhwztetecs replied:
yes, updated!

jcsp commented on src/ceph.in:
```diff
@@ -836,7 +836,7 @@ def get_scrub_timestamps(childargs):
     out = subprocess.check_output(['ceph', 'pg', 'dump', '--format=json-pretty'],
                                   stderr=devnull)
-    for stat in json.loads(out)['pg_stats']:
+    for stat in json.loads(out)['pg_map']['pg_stats']:
```
This will also need to handle the old format (catch KeyError, etc), as our tools are backward compatible by two major versions.

renhwztetecs replied:
updated it.
thanks

tchaikov commented:
this change breaks `rados list-inconsistent-pg`

```
2018-09-19T18:08:02.663 INFO:teuthology.orchestra.run.mira037:Running: 'rados list-inconsistent-pg rbd --format=json'
2018-09-19T18:08:02.748 INFO:teuthology.orchestra.run.mira037.stderr:2018-09-19 18:08:02.740 7ff290d80c40 -1 WARNING: all dangerous and experimental features are enabled.
2018-09-19T18:08:02.784 INFO:teuthology.orchestra.run.mira037.stderr:2018-09-19 18:08:02.776 7ff290d80c40 -1 WARNING: all dangerous and experimental features are enabled.
2018-09-19T18:08:02.817 INFO:teuthology.orchestra.run.mira037.stderr:2018-09-19 18:08:02.809 7ff290d80c40 -1 WARNING: all dangerous and experimental features are enabled.
2018-09-19T18:08:02.826 INFO:teuthology.orchestra.run.mira037.stderr:error 22: (22) Invalid argument
```

see http://pulpito.ceph.com/kchai-2018-09-19_14:35:21-rados-wip-kefu-testing-2018-09-19-1821-distro-basic-mira/3042770/

to be specific, you need to update `get_inconsistent_pgs()` in `librados.cc` , so it can be compatible with both old and new schemes.

renhwztetecs commented:
@tchaikov 
 Added it.  add compatible code to get_inconsistent_pgs() for PG new pg_state_ready field.
Thank you.

tchaikov commented on src/librados/librados.cc:
```diff
@@ -2701,10 +2701,22 @@ namespace {
+      JSONObj *pgstat_obj = parser.find_obj("pg_stats");
+      if (NULL == pgstat_obj)
+        return 0;
+      string s = pgstat_obj->get_data();
```
why would you want to parse it again? instead, i think you could just put
```c++
if (parser.is_array()) {
  v = parser.get_array_elements();
} else if (auto pgstats = parser.find_obj("pg_stats"); pgstats) {
  v = pg_stats.get_array_elements();
} else {
  return -EINVAL;
}
```

tchaikov commented on src/librados/librados.cc:
```diff
@@ -2701,10 +2701,22 @@ namespace {
     if (!parser.is_array()) {
-      return -EINVAL;
+      JSONObj *pgstat_obj = parser.find_obj("pg_stats");
+      if (NULL == pgstat_obj)
```
i think `"pg_stats"` should always exist even it's empty.

tchaikov commented on src/mgr/DaemonServer.cc:
```diff
@@ -1998,6 +1998,16 @@ bool DaemonServer::handle_command(MCommand *m)
+    if (f) {
+       f->open_object_section("pgmap_global");
+       f->open_object_section("pgmap_ready");
+       f->dump_bool("pgmap_ready", pgmap_ready);
```
i don't think we need to enclose "pgmap_ready" with an object. in json, it should be a boolean property of its enclosing object.

i don't really understand what "global" means.

and "pgmap_ready" is a little bit redundant in this context. "ready" would suffice.

renhwztetecs replied:
Yes, ”pgmap_ready” is bool type and does not need to be set to object, updated
other, "pgmap_global" changed to “pg_info” and  "pgmap_ready" changed to "pg_ready"

The purpose of preserving the pg prefix is to match “pg_stats”, In addition, these attributes belong to pg dump.

thank you

renhwztetecs commented:
jenkins retest this please

jcsp commented:
@tchaikov no outstanding comments, right?

liewegas commented:
this broke one of the tests, see http://tracker.ceph.com/issues/36418

liewegas commented:
another bug caused by this?  http://tracker.ceph.com/issues/37444
